### PR TITLE
Image Classifier

### PR DESCRIPTION
**Pull request by ayush-1506**
Should fix: #31

### CODE REVIEW

**Comment by ayush-1506** [comment]
Looks to be working well on the mnist dataset:
```
[ Info: Loss is 0.4725
[ Info: Loss is 0.3228
[ Info: Loss is 0.1755
[ Info: Loss is 0.1288
[ Info: Loss is 0.0445
[ Info: Loss is 0.08269
[ Info: Loss is 0.06347
[ Info: Loss is 0.05062
[ Info: Loss is 0.01943
[ Info: Loss is 0.01023
```

**Comment by ablaom** [comment]
Is the out-of-sample loss on MNIST reasonable?

**Comment by ayush-1506** [comment]
Yes it does seem pretty good. Plus upon further training, the predictions of this trained model also seem to be accurate.

**Comment by ayush-1506** [comment]
@ablaom Would love a review.

**Comment by ablaom** [comment]
The 1.0 test fails because there is a struct defined outside of global scope in test/image.jl . You just need to move the builder definition there to before the @testset block begins.

**Comment by ayush-1506** [comment]
@ablaom Thanks for that, all tests pass now.

**Comment by ablaom** [comment]
@ayush-1506 Many thanks for this. 

There seems to be some deviation from what I recall was the the previously discussed design. Let me say what I recall, and what it's implications are for the implementation, and then you say if you think this is still a good idea. Maybe you are having second thoughts?

My understanding is that the MLJ user does not supply the images `X` as vectors of 3d arrays.  Rather, he provides a vector of images, where a single image is a **matrix** (2d array, not 3d array) whose eltype subtypes `ColorTypes.Gray` (for grey images) or `ColorTypes.AbstractRGB` (for color images). I seem to remember agreeing this was going to be the most convenient format for the casual user.

(In the grey case, this corresponds to the scientific type requirement

```julia
scitype(X) <: AbstractVector{<:AbstractMatrix{<:GrayImage}}
```

Here `GrayImage` is a type defined in ScientificTypes.)

So:

- the input in your tests should be of this form

- the `reformat`/`get` methods should be defined to ensure that `collate` method always transforms data `(X, y)` with `X` as above into a Flux-friendly format

-  like fit, the predict method must also apply collate (to `Xnew`) to ensure the flux model can be applied to it

None of the above appear to be implemented in the current PR.

**Comment by ayush-1506** [comment]
@ablaom Yes, I'm having second thoughts towards this approach. After browsing through the MLDatasets package (which is by far the most common source for quickly and efficiently loading image datasets), I feel it's better to get rid of the relying upon ColorTypes and start treating images as straight-forward 3D/4D tensors. For example, when we load MNIST using MLDatasets using `train_images, train_labels = MNIST.traindata()` , train_images is a dense matrix of type `28×28×60000 reinterpret(FixedPointNumbers.Normed{UInt8,8}, ::Array{UInt8,3}):`.  

Similarly, CIFAR10 (which has color images) train data is `32×32×3×50000 reinterpret(FixedPointNumbers.Normed{UInt8,8}, ::Array{UInt8,4})`. 

In summary, what I mean to say is that I couldn't find a lot of cases when we use types provided by ColorTypes.jl, except for the MNIST dataset that comes with Flux. Naturally, we won't prefer Flux for data loading, since we already have MLDatasets specifically for that.

Let me know your thoughts.

**Comment by ablaom** [comment]
Thanks for this input. 

I have given this some thought and also had another look at
ColorTypes.jl. I think this package is well-thought out and I see it
is actively maintained. It lays out clear path for dealing with images in a Julian way. It has been adopted by several other packages in the graphics ecosystem. 
 I don't know any other package that does this. So, I think we should support the format. 

However,
we must also address the valid point you raise about other formats being
ubiquitous and try to make the life of the user, especially the casual
user, as easy as possible.

Images are messy. Let's establish some terminology for
discussion. (Please correct me if you think I am misunderstanding
anything here.)

**channel type** 

Let `F` denote some subtype of `AbstractFloat` that
is sufficiently encompassing of the scalar types used to encode the
level of a single color channel. For example, I guess
`reinterpret(FixedPointNumbers.Normed{UInt8,8}` would need to be a
subtype of `F`? One question I have is how small can `F` be without
ruling out any use-cases?

**single b/w images**

- `_2D = AbstractArray{<:F, 2}` with W x H interpretation

**single color images and some b/w images**

- `_3D = AbstractArray{<:F, 3}` with  W x H x C interpretation (C = 1 for b/w)

**alternative for single color or b/w image**

- `_2d{C} =AbstractArray{C}` where `C<:Union{AbstractRBG, Gray}` where
  `AbstractRGB` and `Gray` come from `ColorTypes`

**image collections**

- `AbstractVector{<:_2D}` (b/w only)
- `AbstractVector{<:_3D}` 
- `AbstractVector{<:_2d{<:C}}`
- `_3Dcollection = AbstractArray{<:F, 3}` (b/w only) with W x H x N interpreration
- `_4Dcollection = AbstractArray{<:F, 4}` with W x H x C x N interpretation

Note in passing, that `_3D` (color images) and `_2Dcollections` are
the same (which leads to issues raised in 2 below).

I suggest that, moving forward, we consider these as options:

1. Our models support only `_2d{Gray}` and `_2d{AbstractRGB}`. This
   means retaining the existing [scitype
   definitions](https://github.com/alan-turing-institute/MLJScientificTypes.jl/blob/b65844b14a0335ef1f936e39e07b0036acd75cb7/src/convention/scitype.jl#L7)
   as they are. However, we add `coerce` methods to MLJScientificTypes
   to carry out various conversions:
   
   - **single images** (`coerce(::Type{<:Image}, ...)`):
       - `_2D` -> `_2d{Gray{T}}` where `T` is `Float32` or maybe some `FixedPoint` type??
	   - `_3D` -> `_2d{RGB{T}}` ditto 
    
   - **collections** (`coerce(::Type{<:AbstractVector{<:Image}, ...)`):
       - `_3Dcollection` -> `Vector{_2d{Gray{T}}}`
       - `_4Dcollection` -> `Vector{_2d{C}}` where `C` is `Gray{T}` or `RGB{T}`. 


2. Our models support all of the image collection types. This would
   require adding:

    - scitype definitions to declare any object subtying
      `_2D`, `_3D` as `ColorImage` or `GrayImage` appropriately
    - adding a new scitype `ImageCollection` and writing a scitype method
       to decare the concrete collections to have that scitype. 

    We wouldn't need any `coerce` method additions.
   
Option 2 has some issues: (i) It is conceivable that objects of type
`_2D`, `_3D`, `_3Dcollection` and/or `_4Dcollection` are used in a
non-image context (`_2d` never will). There has been some debate, for
example, whether supervised models that take tabular data with float
data should not also be allowed to accept accept float matrices, which
would have type `_2D`. (ii) In implementing the new scitype methods,
how would we distinguish between `_3D` objects (scitype `<:Image`) and
`_2Dcollections` (scitype `ImageCollections`), since these are
actually the same machine types?

The only disadvantage of 1 that is immediately obvious to me is that
users will sometimes have to first perform a `coerce` (and they have to
*know* that this is what they should do). However, type coercions are
pretty unavoidable in general, and (hopefully) the scitype formalism
has made this a little easier for users.

I expect we can only overcome the issues raised in 2 by adding some finicky
complications. Somehow option 1 seems simpler and safer than trying to
catch all possible types at the level of the model API.
	   
Thoughts?

**Comment by ablaom** [comment]
@ayush Okay, looks like you made a lot of progress!

I fixed a few things with the Project file to get access to the latest MLJScientific types. And I fixed some  minor issues with the `reformat` method. And I fixed the first image test, which was not updated to reflect the new API. 

There is still a problem with the image `predict` method. I had a go at it but ran out of time. Perhaps you can revisit it. Keep in mind that `Xnew` is essentially the same kind of object that as the `X` received by `fit`.

**Comment by ayush-1506** [comment]
@ablaom Thanks for that. I've pushed a commit that should get `predict` working.

**Comment by ayush-1506** [comment]
Hi @ablaom  just a reminder, would love another review. Thanks

**Comment by ayush-1506** [comment]
Hi @ablaom just a reminder, would love another review. Thanks

**Comment by ablaom** [comment]
Thanks for your patience. I have been catching up after a short break and hope to get to this tomorrow. In the meantime, if you have chance, would be great if you can fix the julia 1.0 fail (`range` issue).

**Comment by ablaom** [comment]
Thanks! 

I added a test for batch_size > 1, which fails. Be great if you can investigate.

**Comment by ayush-1506** [comment]
There was an error in the chain, fixed it. Note to self: not a good idea to use `vec` inside a Flux chain.

**Comment by ayush-1506** [comment]
@ablaom Test fails because loss doesn't decrease by 0.8 times. I guess that's because the test uses random images and targets, so the loss decreasing probably doesn't make a lot of sense?
